autoprecompiles/src/optimizer.rs

```diff
@@ -282,31 +282,6 @@ fn bus_interaction_to_symbolic_bus_interaction<P: FieldElement>(
     }
 }
 
-/// Turns an algebraic expression into a quadratic symbolic expression,
-/// assuming all [`AlgebraicReference`]s are unknown variables.
-pub fn algebraic_to_quadratic_symbolic_expression<T: FieldElement>(
-    expr: &AlgebraicExpression<T>,
-) -> QuadraticSymbolicExpression<T, AlgebraicReference> {
-    powdr_expression::conversion::convert(expr, &mut |reference| {
-        QuadraticSymbolicExpression::from_unknown_variable(reference.clone())
-    })
-}
-
-/// Turns a quadratic symbolic expression back into an algebraic expression.
-/// Tries to simplify the expression wrt negation and constant factors
-/// to aid human readability.
-pub fn quadratic_symbolic_expression_to_algebraic<T: FieldElement>(
-    expr: &QuadraticSymbolicExpression<T, AlgebraicReference>,
-) -> AlgebraicExpression<T> {
-    // Wrap `powdr_pilopt::qse_opt::quadratic_symbolic_expression_to_algebraic`, which
-    // works on a `powdr_ast::analyzed::AlgebraicExpression`.
-    let expr = expr.transform_var_type(&mut |algebraic_reference| {
-        powdr_pilopt::qse_opt::Variable::Reference(algebraic_reference.clone().into())
-    });
-    // This is where the core conversion is implemented, including the simplification.
-    let ast_algebraic_expression =
-        powdr_pilopt::qse_opt::quadratic_symbolic_expression_to_algebraic(&expr);
-    // Unwrap should be fine, because by construction we don't have challenges or public references,
-    // and quadratic_symbolic_expression_to_algebraic should not introduce any exponentiations.
-    AlgebraicExpression::try_from_ast_expression(ast_algebraic_expression).unwrap()
+pub fn simplify_expression<T: FieldElement>(e: AlgebraicExpression<T>) -> AlgebraicExpression<T> {
+    quadratic_symbolic_expression_to_algebraic(&algebraic_to_quadratic_symbolic_expression(&e))
```

Comment:
This is the main change in this PR. I don't know the details of pilopt's `simplify_expression`, but this should also achieve a good amount of simplification.

Reply:
I'm not sure I understand, in which crate would the implementation live? Because the `expression` crate doesn't know `constraint_system` (where QSE lives) and vice versa.

I think this discussion is also orthogonal to this PR.

Reply:
Ah right they're both does, fair enough, closing